runtime/controller.go

```diff
@@ -115,11 +117,41 @@ func NewController(rt *Runtime, instanceID string, logger *zap.Logger, ac activi
 		logger = logger.With(zap.String("instance_id", instanceID))
 		logger = logger.Named("console")
 	}
-	c.Logger = slog.New(zapslog.HandlerOptions{LoggerName: "console"}.New(logger.Core()))
+
+	c.Logs = logbuffer.NewBuffer(rt.opts.ControllerLogBufferCapacity, rt.opts.ControllerLogBufferSizeBytes)
+	c.Logger = slog.New(&duplicatingHandler{
+		zapHandler: zapslog.HandlerOptions{LoggerName: "console"}.New(logger.Core()),
+		logs:       c.Logs,
+	})
 
 	return c, nil
 }
 
+type duplicatingHandler struct {
+	zapHandler *zapslog.Handler
+	logs       *logbuffer.Buffer
+}
+
+func (d *duplicatingHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return d.zapHandler.Enabled(ctx, level)
+}
+
+func (d *duplicatingHandler) Handle(ctx context.Context, record slog.Record) error {
+	err := d.zapHandler.Handle(ctx, record)
+	if err != nil {
+		return err
+	}
+	return d.logs.Add(record)
+}
+
+func (d *duplicatingHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	panic("not implemented")
+}
+
+func (d *duplicatingHandler) WithGroup(name string) slog.Handler {
+	panic("not implemented")
+}
```

Comment:
1. We should implement `WithAttrs` and `WithGroup` (it should be quite simple, something like  `return duplicatingHandler{zapHandler: d.zapHandler.With...}`)
2. Should we move the duplicatingHandler into `logbuffer` or another util package? Or alternatively move to the bottom of this file, since it's a util (but this file is already so big...).

Reply:
1. `zapHandler.WithAttrs(attrs)` returns a `slog.Hanlder` not `zapslog.Handler` which is needed here. Also if we implement `WithAttrs` wouldn't we need to store attributes in duplicating handler so that they can be added to the payload of the `Log` in the buffer.
2. Don't think it belongs to the `logbuffer` package, probably we can rename the package to `logutils` and add it there.

Reply:
1. It seems like one can simply typecast it to `slog.Handler`

Reply:
Okay – I think we should still keep the `logbuffer` package for the log buffer since it seems like a self-contained thing. We can add a `logutil` package for the duplicating handler.

A thought: do you think all this would be easier if we didn't use `slog` and just used `zap` with a custom backend?

Reply:
I think that will involve implementing `zapcore.Core` and use `zapcore.Tee`, will have to implement and see how easy it is